File: scripts/generate_3d_lobsters_meshy.py
import os
import json
import time
import base64
import urllib.request
import urllib.error

API_KEY = os.environ.get("MESHY_API_KEY", "msy_dummy_api_key_for_test_mode_12345678")
HEADERS = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json"
}

LOBSTERS = ["FlatIvyBase.png"]
SOURCE_DIR = os.path.join(os.path.dirname(__file__), "../public/agents")
TARGET_DIR = os.path.join(os.path.dirname(__file__), "../public/models/lobsters")

if not os.path.exists(TARGET_DIR):
    os.makedirs(TARGET_DIR)

def image_to_base64_uri(file_path):
    with open(file_path, "rb") as f:
        data = f.read()
    b64 = base64.b64encode(data).decode('utf-8')
    return f"data:image/png;base64,{b64}"

def start_meshy_task(image_url):
    req = urllib.request.Request(
        "https://api.meshy.ai/openapi/v1/image-to-3d",
        headers=HEADERS,
        data=json.dumps({
            "image_url": image_url,
            "enable_pbr": True,
            "topology": "quad",
            "target_polycount": 30000
        }).encode('utf-8')
    )
    try:
        with urllib.request.urlopen(req) as response:
            result = json.loads(response.read())
            return result.get("result")
    except urllib.error.HTTPError as e:
        print(f"HTTPError: {e.code} - {e.read().decode('utf-8')}")
        raise

def poll_meshy_task(task_id, file_name):
    url = f"https://api.meshy.ai/openapi/v1/image-to-3d/{task_id}"
    req = urllib.request.Request(url, headers=HEADERS)
    attempt = 0
    while attempt < 120:
        attempt += 1
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read())
        print(f"[{file_name}] Status: {data.get('status')} (Progress: {data.get('progress')}%)")
        if data.get("status") == "SUCCEEDED":
            return data["model_urls"]["glb"]
        elif data.get("status") in ("FAILED", "EXPIRED"):
            raise Exception(f"Task failed: {data.get('task_error')}")
        time.sleep(10)
    raise Exception("Polling timeout")

def download_glb(url, file_name):
    dest = os.path.join(TARGET_DIR, file_name.replace(".png", ".glb"))
    urllib.request.urlretrieve(url, dest)
    print(f"Downloaded to {dest}")

def process(lobster):
    source = os.path.join(SOURCE_DIR, lobster)
    if not os.path.exists(source):
        print(f"Skipping {lobster}, file not found")
        return
    print(f"Processing {lobster}...")
    try:
        b64 = image_to_base64_uri(source)
        task_id = start_meshy_task(b64)
        print(f"Started task {task_id}")
        glb_url = poll_meshy_task(task_id, lobster)
        download_glb(glb_url, lobster)
    except Exception as e:
        print(f"Error processing {lobster}: {e}")

import concurrent.futures

if __name__ == "__main__":
    print(f"Starting Meshy generation for {len(LOBSTERS)} models...")
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        futures = {executor.submit(process, item): item for item in LOBSTERS}
        for future in concurrent.futures.as_completed(futures):
            item = futures[future]
            try:
                future.result()
                print(f"[{item}] Successfully completed.")
            except Exception as exc:
                print(f"[{item}] Generated an exception: {exc}")
    print("All generation attempts completed.")

File: scripts/generate_3d_lobsters_meshy.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Ensure process.env.MESHY_API_KEY is available or use a fallback test key if possible.
const MESHY_API_KEY = process.env.MESHY_API_KEY;

if (!MESHY_API_KEY) {
  console.error("Missing MESHY_API_KEY environment variable. Using test mode via meshy.ai dummy key if it works, otherwise this will fail.");
}

// Fallback to test key if user does not set one (based on web search for test mode keys)
// "msy_dummy_api_key_for_test_mode_12345678"
const API_KEY = MESHY_API_KEY || "msy_dummy_api_key_for_test_mode_12345678";
const HEADERS = {
  "Authorization": `Bearer ${API_KEY}`,
  "Content-Type": "application/json"
};

const LOBSTERS = [
  "Accountant.png",
  "Assistant.png",
  "Strategist.png",
  "Researcher.png",
  "Tutor.png"
];

const SOURCE_DIR = path.join(__dirname, '../public/agents');
const TARGET_DIR = path.join(__dirname, '../public/models/lobsters');

if (!fs.existsSync(TARGET_DIR)) {
  fs.mkdirSync(TARGET_DIR, { recursive: true });
}

function imageToBase64Uri(filePath) {
  const data = fs.readFileSync(filePath);
  const base64 = data.toString('base64');
  return `data:image/png;base64,${base64}`;
}

async function startMeshyTask(imageUrl) {
  const response = await fetch("https://api.meshy.ai/openapi/v1/image-to-3d", {
    method: "POST",
    headers: HEADERS,
    body: JSON.stringify({
      image_url: imageUrl,
      enable_pbr: true,
      topology: "quad",     // Best for characters
      target_polycount: 30000 
    })
  });
  
  const text = await response.text();
  console.log("Raw Meshy response:", text);

  if (!response.ok) {
    throw new Error(`Failed to start task: ${text}`);
  }

  const data = JSON.parse(text);
  return data.result; // task_id
}

async function pollMeshyTask(taskId, fileName) {
  let attempt = 0;
  while (attempt < 120) { // Max wait ~ 20 mins
    attempt++;
    const response = await fetch(`https://api.meshy.ai/openapi/v1/image-to-3d/${taskId}`, {
      headers: HEADERS
    });

    if (!response.ok) {
      throw new Error(`Polling failed: ${await response.text()}`);
    }

    const data = await response.json();
    console.log(`[${fileName}] Task Status: ${data.status} (Progress: ${data.progress}%)`);

    if (data.status === "SUCCEEDED") {
      return data.model_urls.glb;
    } else if (data.status === "FAILED" || data.status === "EXPIRED") {
      throw new Error(`Task failed or expired: ${JSON.stringify(data.task_error)}`);
    }

    // Wait 10 seconds before polling again
    await new Promise(resolve => setTimeout(resolve, 10000));
  }
  throw new Error("Task timeout exceeded.");
}

async function downloadGLB(url, fileName) {
  const dest = path.join(TARGET_DIR, fileName.replace('.png', '.glb'));
  const response = await fetch(url);
  
  if (!response.ok) {
    throw new Error(`Failed to download GLB for ${fileName}`);
  }

  const arrayBuffer = await response.arrayBuffer();
  fs.writeFileSync(dest, Buffer.from(arrayBuffer));
  console.log(`Successfully downloaded: ${dest}`);
}

async function processLobster(fileName) {
  const sourceFile = path.join(SOURCE_DIR, fileName);
  if (!fs.existsSync(sourceFile)) {
    console.warn(`Source file not found: ${fileName}. Skipping.`);
    return;
  }

  try {
    console.log(`Processing ${fileName}... converting to base64...`);
    const base64Uri = imageToBase64Uri(sourceFile);
    
    console.log(`Sending to Meshy API to start generation...`);
    const taskId = await startMeshyTask(base64Uri);
    console.log(`Task created with ID: ${taskId}`);

    console.log(`Polling for completion...`);
    const glbUrl = await pollMeshyTask(taskId, fileName);
    
    console.log(`Model generated! Downloading from ${glbUrl}...`);
    await downloadGLB(glbUrl, fileName);
    
  } catch (error) {
    console.error(`Error processing ${fileName}:`, error.message);
  }
}

async function main() {
  for (const lobster of LOBSTERS) {
    await processLobster(lobster);
  }
  console.log("All generation attempts completed.");
}

main().catch(err => console.error(err));
